### DISCUSSION TITLE
Simulate edge cancellation

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#680] [Q&A]
I often want to be able to cancel an individual task, and have something that looks like asyncio's edge cancellation API, while keeping the advantages of structured concurrency.
In AnyIO this means creating a task group just for that task, but since this operation is async, one has to await the creation of the task group before being able to cancel the task, which is not the case in asyncio where launching and cancelling a task can be done in non-async functions:
```py
from asyncio import create_task, run, sleep

async def foo():
    print("start")
    await sleep(10)
    print("stop")

async def main():
    task = create_task(foo())
    task.cancel()

run(main())
```
The following `Cancellable` class is close to asyncio's API, but cancelling needs to be awaited:
```py
from anyio import Event, create_task_group, run, sleep

class Cancellable:
    def __init__(self, coro):
        self._coro = coro
        self._event = Event()

    async def cancel(self):
        await self._event.wait()
        self._tg.cancel_scope.cancel()

    async def run(self):
        async with create_task_group() as self._tg:
            self._event.set()
            await self._coro

async def foo():
    print("start")
    await sleep(10)
    print("stop")

async def main():
    async with create_task_group() as tg:
        cancellable = Cancellable(foo())
        tg.start_soon(cancellable.run)
        await cancellable.cancel()

run(main)
```
And the following version has a non-async cancellation:
```py
from anyio import Event, create_task_group, run, sleep
from anyio.abc import TaskGroup

class Cancellable:
    def __init__(self, coro):
        self._coro = coro
        self._stop_event = Event()

    def cancel(self):
        self._stop_event.set()

    async def run(self):
        async with create_task_group() as tg:
            tg.start_soon(self._run, tg)
            tg.start_soon(self._check_stop, tg)

    async def _run(self, tg: TaskGroup):
        await self._coro
        tg.cancel_scope.cancel()

    async def _check_stop(self, tg: TaskGroup):
        await self._stop_event.wait()
        tg.cancel_scope.cancel()

async def foo():
    print("start")
    await sleep(10)
    print("stop")

async def main():
    async with create_task_group() as tg:
        cancellable = Cancellable(foo())
        tg.start_soon(cancellable.run)
        cancellable.cancel()

run(main)
```
I'm not sure this is the right way to do it, maybe I'm missing something simpler?
Do you think simulating edge cancellation is not a good idea in the first place?

**Comment by TayfurYldz** [comment]
There is now a simpler built-in way to express this.

Since AnyIO 4.14, both `TaskGroup.start_soon()` and `TaskGroup.create_task()` return a `TaskHandle`. A handle can cancel just that task synchronously, while the task still remains owned by its surrounding task group:

```python
from anyio import create_task_group, run, sleep

async def foo():
    print("start")
    await sleep(10)
    print("stop")

async def main():
    async with create_task_group() as tg:
        handle = tg.start_soon(foo)
        handle.cancel()

run(main)
```

So for the use case in the question, there is no longer a need to create a one-task nested task group, wait for an event exposing its cancel scope, or run a second watcher task solely to make `cancel()` synchronous.

This keeps the structured-concurrency property that matters here: the child still belongs to the lexical lifetime of `tg`, and exiting the task group still waits for/collects its children. The handle is just a way to address one child rather than cancelling the task group's whole cancel scope.

It is worth distinguishing this from trying to reproduce asyncio's cancellation semantics globally. AnyIO still uses level cancellation through cancel scopes, and code should remain cancellation-safe under those semantics. `TaskHandle.cancel()` solves the *individual task ownership/control* problem without requiring the application to emulate asyncio's edge-cancellation model.
